src/Access/RoleCache.h

```diff
@@ -27,9 +27,13 @@ class RoleCache
     void collectEnabledRoles(scope_guard * notifications);
     void collectEnabledRoles(EnabledRoles & enabled, scope_guard * notifications);
     RolePtr getRole(const UUID & role_id);
+    RolePtr makeRoleEntry(const UUID & role_id);
+
     void roleChanged(const UUID & role_id, const RolePtr & changed_role);
     void roleRemoved(const UUID & role_id);
 
+    bool isRoleEnabled(const UUID & role_id);
+
     const AccessControl & access_control;
```

Comment:
The following scenario:
1. A user `A` with an attached role `AR` logs in
2. 10 minutes pass
3. Another user `B` with another attached role `BR` logs in, purging the role `AR` from the cache `RoleCache::cache`.
4. An administrator grant something to `AR`. But the subscription to `AR` is lost (see 4) so the access rights won't be updated.

This is the reason why I suggested to keep subscriptions not only in `cache` but also somewhere near `enabled_roles`.

Reply:
@vitlibar , yes, sure.
Currently, only the simplest case with one role is covered.
When I am done with development I'll add another test.
I have in mind a slightly different scenario.
1.  == " ==
2.  == " ==
3. Another role ZR is altered (and AR is purged from the cache therefore)
4.  == " ==
But effectively it is the same, right?

Reply:
It looks like we're talking about different fixes. Ok, I'll do my fix and you can do yours.

Reply:
> It looks like we're talking about different fixes. Ok, I'll do my fix and you can do yours.

Why do you think so?

From my point of view it is the same.

I've pushed current state of my work.